### PR TITLE
Add mesher

### PR DESCRIPTION
**Pull request by sinaatalay**
This PR adds basic 2D mesh generation capabilities.

### CODE REVIEW

**Comment by sinaatalay** [comment]
> Would it be too difficult to add a demo script before my review?

I added some new things. Can you try the code below?


```python
import fastfem.mesh as m

mesh = m.Square(
    side_length=1.0,
    nodes_in_horizontal_direction=10,
    nodes_in_vertical_direction=10,
    element_type="triangle",
).mesh
```

File: pyproject.toml [hunk 1]
```diff
@@ -131,21 +139,6 @@ build = "mkdocs build --clean --strict" # hatch run docs:build
...
-    "time-machine==2.15.0", # to select an arbitrary date and time for testing
-    "pypdf==4.3.1",         # to read PDF files
-]
-path = ".venv-tests"
-
-[tool.hatch.envs.tests.scripts]
-# Run the tests:
-run = "pytest" # hatch run tests:run  
-# Run the tests and generate the coverage report as HTML:
-run-and-report = "coverage run -m pytest && coverage report && coverage html --show-contexts" # hatch run tests:coverage
```
Comment (int-ptr-ptr) [inline pyproject.toml:148]: Is there a particular reason for removing `.venv-tests`?

**Review by int-ptr-ptr — CHANGES_REQUESTED** [review]
You're not doing any sort of garbage collection with gmsh. Running the square code you provided will properly generate the mesh, but changing `nodes_in_<horizontal/vertical>_direction` and running the code again will not give a new mesh with the desired values.

Additionally, it would be helpful to have more members than just the generated mesh dictionary you gave. Here is an example use case that could use other fields:

```python
import numpy as np
import matplotlib.pyplot as plt

import fastfem.mesh as m
square = m.Square(
    side_length=2.0,
    nodes_in_horizontal_direction=None,
    nodes_in_vertical_direction=None,
    element_type="triangle",
)
mesh = square.mesh


num_nodes_full = max(
    max(
        max(meshentry.node_tags)
        for meshentry in meshlist
    )
    for meshlist in mesh.values()
) #I shouldn't need to do this
COORD_DIM = 3

node_coords_full = np.empty((num_nodes_full,COORD_DIM))
was_node_set = np.zeros(num_nodes_full,dtype=bool)


surface_meshes = []
volume_meshes = []
#I use the convention of volumetric = N dim, and surface = (N-1) dim

for name,tag in mesh.keys():
    # just a personal gripe that I don't really like dictionaries for things that don't
    # change much (after initial creation) since you don't utilize the O(1) insert, while
    # still paying the memory cost of unsaturated table space. O(1) access can be achieved
    # arrays by having workers store indices that they will need.
    # That being said, we can keep it a dict if you prefer it.

    meshlist = mesh[(name,tag)]
    print(f"{name}({tag}) -- #entries: {len(meshlist)}") #will this ever be > 1?
    for imesh,meshentry in enumerate(meshlist):

        #these should probably be member attributes
        num_nodes = len(meshentry.node_tags)
        num_elems = len(meshentry.element_tags)
        nodes_per_elem = len(meshentry.nodes_of_elements)//num_elems
        #this will fail for different element types, I think.
        element_type = "surface" if nodes_per_elem == 2 else "volumetric"

        assert len(meshentry.nodes_of_elements) % num_elems == 0


        print(f"    {name}[{imesh}]:")
        print(f"        type: {element_type}")
        print(f"        number of elements: {num_elems}")
        print(f"        number of nodes: {num_nodes}")
        if element_type == "surface":
            surface_meshes.append((name,imesh,meshentry))
        else:
            volume_meshes.append((name,imesh,meshentry))
        node_tags = meshentry.node_tags - 1 # This off-by-one error from indexing :/
        node_coords = meshentry.coordinates_of_nodes.reshape((num_nodes,COORD_DIM)) # < I shouldn't need to do a reshape


        #to verify tagged nodes are the same, pull out the global array
        prev_set = was_node_set[node_tags]
        prev_vals = node_coords_full[node_tags,:]

        #compare global array's set values with the replace values
        assert np.max(np.linalg.norm(prev_vals[prev_set,:] - node_coords[prev_set,:])) < 1e-8

        #set global array
        node_coords_full[node_tags] = node_coords
        was_node_set[node_tags] = True



#we should have received coords for all nodes
assert np.all(was_node_set)



plt.figure(0,figsize=(10,6))
colors = "rygcbmk"
last_color = 0
legend_handles = []
#graph volumetric elemnts first
for name,imesh,meshentry in volume_meshes:
    color = colors[last_color]
    last_color = (last_color + 1) % len(colors)
    style = ":"+color
    #transpose for plotting

    num_elems = len(meshentry.element_tags)
    nodes_per_elem = len(meshentry.nodes_of_elements)//num_elems
    elem_nodes = meshentry.nodes_of_elements.reshape((num_elems,nodes_per_elem)) - 1
    elem_coords = node_coords_full[elem_nodes.T,:]
    
    #elem_wrap just closes the polygon (append fist node of each element)
    if nodes_per_elem > 2:
        elem_wrap = np.concatenate([elem_coords,elem_coords[0,np.newaxis,:,:]],axis=0)
    else:
        elem_wrap = elem_coords
    handles = plt.plot(elem_wrap[:,:,0],elem_wrap[:,:,1],style,label=f"{name}[{imesh}]")
    legend_handles.append(handles[0])

#then graph boundaries
for name,imesh,meshentry in surface_meshes:
    color = colors[last_color]
    last_color = (last_color + 1) % len(colors)
    style = "-."+color
    #transpose for plotting

    num_elems = len(meshentry.element_tags)
    nodes_per_elem = len(meshentry.nodes_of_elements)//num_elems
    elem_nodes = meshentry.nodes_of_elements.reshape((num_elems,nodes_per_elem)) - 1
    elem_coords = node_coords_full[elem_nodes.T,:]
    
    #elem_wrap just closes the polygon (append fist node of each element)
    if nodes_per_elem > 2:
        elem_wrap = np.concatenate([elem_coords,elem_coords[0,np.newaxis,:,:]],axis=0)
    else:
        elem_wrap = elem_coords
    handles = plt.plot(elem_wrap[:,:,0],elem_wrap[:,:,1],style,label=f"{name}[{imesh}]")
    legend_handles.append(handles[0])

plt.scatter(node_coords_full[:,0],node_coords_full[:,1])
plt.legend(handles=legend_handles)
plt.title("Mesh Output")
plt.show()
```

image

**Comment by int-ptr-ptr** [comment]
Here's a test that currently fails:

```python
def test_two_instances():
    m.Square(
        side_length=1.0,
        nodes_in_horizontal_direction=None,
        nodes_in_vertical_direction=None,
        element_type="triangle",
    )
    m.Square(
        side_length=1.0,
        nodes_in_horizontal_direction=5,
        nodes_in_vertical_direction=5,
        element_type="triangle",
    )
```

File: pyproject.toml (hunk 1 above)
Comment (sinaatalay) [reply pyproject.toml:148]: Yes, I moved test dependencies it to `.venv`. Keeping two virtual environments for development and tests was very hard to use; you had to switch between them all the time. Usually, testing and development take place at the same time. That's why I did it.

**Comment by sinaatalay** [comment]
Hi, I have made some changes, no more garbage collection problem.

Could you try running the code below?

```py
import fastfem.mesh as m

mesh = m.create_a_rectangle_mesh(
    horizontal_length=1,
    vertical_length=1,
    nodes_in_horizontal_direction=10,
    nodes_in_vertical_direction=10,
    element_type="quadrangle",
    file_name=None,
)

bottom_boundary = mesh["bottom_boundary"]
right_boundary = mesh["right_boundary"]
top_boundary = mesh["top_boundary"]
left_boundary = mesh["left_boundary"]
rectangle = mesh["rectangle"]

# The following attributes are available for the each domain:
dimension = bottom_boundary.dimension
for submesh in bottom_boundary.mesh:
    submesh.node_tags
    submesh.coordinates_of_nodes
    submesh.element_tags
    submesh.nodes_of_elements
    submesh.type
```

Also, could you please tell me what other information about the mesh is needed? From your comment above, I didn't understand what you wanted me to add.

**Comment by int-ptr-ptr** [comment]
I am assuming we are using one-indexing. I do have my reservations about that in a zero-indexed language, but it should be fine.

> Also, could you please tell me what other information about the mesh is needed? From your comment above, I didn't understand what you wanted me to add.

Sure, I can list out the features that I think are a must:
- [ ] `Mesh.num_nodes` -- total number of nodes, should be equal to `max(max(max(meshentry.node_tags) for meshentry in domain.mesh) for domain in mesh)`, which gives the largest node tag in every submesh. Alternatively, one could use the shape of the array below:
- [ ] `Mesh.nodes` -- a `Mesh.num_nodes ✕ 3` array of node locations. It doesn't make sense to me that node coordinates are all packed away in each submesh. `Submesh.coordinates_of_nodes` can be replaced with views of `Mesh.nodes` (i.e. `Mesh.nodes[Submesh.node_tags-1,:]`)
- [ ] Add `Submesh.nodes_per_element` and or reshape `Submesh.nodes_of_elements` to be shape `(len(Submesh.element_tags),nodes_per_element)`
- [ ] Create some sort of iteration scheme of volume and surface submeshes and/or elements (like the `volume_meshes` and `surface_meshes` in my example).

These aren't as necessary, but I think would be nice to have:
- [ ] `Submesh.num_nodes = len(Submesh.node_tags)`
- [ ] `Submesh.num_elements = len(Submesh.element_tags)`
- [ ] `Mesh.__getitem__(self, key: str)` -> `Mesh.__getitem__(self, key: str | int)` for tag-indexing
- [ ] `Mesh.domain_names = [domain.name for domain in Mesh]`
- [ ] `Mesh.domain_tags = [domain.tag for domain in Mesh]`

**Comment by int-ptr-ptr** [comment]
<details>

<summary>Here is an updated version of the code from a prior comment.</summary>

```python
import numpy as np
import matplotlib.pyplot as plt
import fastfem.mesh as m

square = m.create_a_square_mesh(
    side_length=2.0,
    # nodes_in_horizontal_direction=10,
    # nodes_in_vertical_direction=10,
    element_type="triangle",
)
mesh = square


num_nodes_full = max(
    max(
        max(meshentry.node_tags)
        for meshentry in domain.mesh
    )
    for domain in mesh
) #I shouldn't need to do this
COORD_DIM = 3

node_coords_full = np.empty((num_nodes_full,COORD_DIM))
was_node_set = np.zeros(num_nodes_full,dtype=bool)


surface_meshes = []
volume_meshes = []
#I use the convention of volumetric = N dim, and surface = (N-1) dim

for domain in mesh:
    name = domain.name
    # just a personal gripe that I don't really like dictionaries for things that don't
    # change much (after initial creation) since you don't utilize the O(1) insert, while
    # still paying the memory cost of unsaturated table space. O(1) access can be achieved
    # arrays by having workers store indices that they will need.
    # That being said, we can keep it a dict if you prefer it.

    
    print(f"{name}({domain.tag}) -- {domain.dimension}D, #entries: {len(domain.mesh)}") #will this ever be > 1?
    for imesh,meshentry in enumerate(domain.mesh):

        #these should probably be member attributes
        num_nodes = len(meshentry.node_tags)
        num_elems = len(meshentry.element_tags)
        nodes_per_elem = len(meshentry.nodes_of_elements)//num_elems
        #this will fail for different element types, I think.
        element_type = "surface" if nodes_per_elem == 2 else "volumetric"

        assert len(meshentry.nodes_of_elements) % num_elems == 0


        print(f"    {name}[{imesh}]:")
        print(f"        type: {element_type}")
        print(f"        number of elements: {num_elems}")
        print(f"        number of nodes: {num_nodes}")
        if element_type == "surface":
            surface_meshes.append((name,imesh,meshentry))
        else:
            volume_meshes.append((name,imesh,meshentry))
        node_tags = meshentry.node_tags - 1 # This off-by-one error from indexing :/
        node_coords = meshentry.coordinates_of_nodes.reshape((num_nodes,COORD_DIM)) # < I shouldn't need to do a reshape


        #to verify tagged nodes are the same, pull out the global array
        prev_set = was_node_set[node_tags]
        prev_vals = node_coords_full[node_tags,:]

        #compare global array's set values with the replace values
        assert np.max(np.linalg.norm(prev_vals[prev_set,:] - node_coords[prev_set,:])) < 1e-8

        #set global array
        node_coords_full[node_tags] = node_coords
        was_node_set[node_tags] = True



#we should have received coords for all nodes
assert np.all(was_node_set)



plt.figure(0,figsize=(10,6))
colors = "rygcbmk"
last_color = 0
legend_handles = []
#graph volumetric elemnts first
for name,imesh,meshentry in volume_meshes:
    color = colors[last_color]
    last_color = (last_color + 1) % len(colors)
    style = ":"+color
    #transpose for plotting

    num_elems = len(meshentry.element_tags)
    nodes_per_elem = len(meshentry.nodes_of_elements)//num_elems
    elem_nodes = meshentry.nodes_of_elements.reshape((num_elems,nodes_per_elem)) - 1
    elem_coords = node_coords_full[elem_nodes.T,:]
    
    #elem_wrap just closes the polygon (append fist node of each element)
    if nodes_per_elem > 2:
        elem_wrap = np.concatenate([elem_coords,elem_coords[0,np.newaxis,:,:]],axis=0)
    else:
        elem_wrap = elem_coords
    handles = plt.plot(elem_wrap[:,:,0],elem_wrap[:,:,1],style,label=f"{name}[{imesh}]")
    legend_handles.append(handles[0])

#then graph boundaries
for name,imesh,meshentry in surface_meshes:
    color = colors[last_color]
    last_color = (last_color + 1) % len(colors)
    style = "-."+color
    #transpose for plotting

    num_elems = len(meshentry.element_tags)
    nodes_per_elem = len(meshentry.nodes_of_elements)//num_elems
    elem_nodes = meshentry.nodes_of_elements.reshape((num_elems,nodes_per_elem)) - 1
    elem_coords = node_coords_full[elem_nodes.T,:]
    
    #elem_wrap just closes the polygon (append fist node of each element)
    if nodes_per_elem > 2:
        elem_wrap = np.concatenate([elem_coords,elem_coords[0,np.newaxis,:,:]],axis=0)
    else:
        elem_wrap = elem_coords
    handles = plt.plot(elem_wrap[:,:,0],elem_wrap[:,:,1],style,label=f"{name}[{imesh}]")
    legend_handles.append(handles[0])

plt.scatter(node_coords_full[:,0],node_coords_full[:,1])
plt.legend(handles=legend_handles)
plt.title("Mesh Output")
plt.show()
```

</details>

**Comment by sinaatalay** [comment]
I have finished the mesher. Please let me know if you need anything else @int-ptr-ptr. Other than that, it's ready for the review @sacha-escudier.

The API has been changed a bit. Here is a sample script for the new API:

```py
import fastfem.mesh as m

mesh = m.create_a_rectangle_mesh(
    horizontal_length=1,
    vertical_length=1,
    nodes_in_horizontal_direction=10,
    nodes_in_vertical_direction=10,
    element_type="quadrangle",
    file_name=None,
)

# Dictionary of all nodes. Keys are tags, values are coordinates.
mesh.nodes

# Number of nodes:
mesh.number_of_nodes

# Domains:
bottom_boundary = mesh["bottom_boundary"]
right_boundary = mesh["right_boundary"]
top_boundary = mesh["top_boundary"]
left_boundary = mesh["left_boundary"]
rectangle = mesh["surface"]

# The following attributes are available for the each domain:
dimension = bottom_boundary.dimension
for submesh in bottom_boundary.mesh:
    submesh.nodes
    # Dictionary of all elements. Keys are tags, values are node tags.
    submesh.elements

    submesh.number_of_nodes_per_element
    submesh.number_of_elements
```

**Review by int-ptr-ptr — APPROVED** [review]
Looks good!
Thank you!

**Comment by sacha-escudier** [comment]
> # The following attributes are available for the each domain:
> dimension = bottom_boundary.dimension
> for submesh in bottom_boundary.mesh:
>     submesh.node_tags
>     submesh.coordinates_of_nodes
>     submesh.element_tags
>     submesh.nodes_of_elements
>     submesh.type

I may just be missing something, but you previously had a way of accessing the element type of a mesh through the attribute mesh.type. Is it still available with the newest version you provided? Or has the attribute name changed?

**Comment by sinaatalay** [comment]
@sacha-escudier `type` is still there. Your IDE should be able to show it. Let me know if you can't access it.

**Review by sacha-escudier — APPROVED** [review]
Looks good, thanks!

**Pull request merged**